core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java

```diff
@@ -14,133 +14,384 @@
 import alluxio.Configuration;
 import alluxio.PropertyKey;
 import alluxio.master.journal.Journal;
-import alluxio.master.journal.JournalFormatter;
 import alluxio.master.journal.JournalReader;
+import alluxio.master.journal.JournalWriter;
+import alluxio.master.journal.options.JournalReaderCreateOptions;
+import alluxio.master.journal.options.JournalWriterCreateOptions;
 import alluxio.underfs.UnderFileStatus;
 import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.DeleteOptions;
 import alluxio.util.URIUtils;
+import alluxio.util.UnderFileSystemUtils;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.net.URI;
-import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
 
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
  * Implementation of UFS-based journal.
  *
  * The journal is made up of 2 components:
- * - The checkpoint: the full state of the master
- * - The entries: incremental entries to apply to the checkpoint.
+ * - The checkpoint: the full state of the master.
+ * - The log entries: incremental entries to apply to the checkpoint.
+ *
+ * The journal log entries must be self-contained. Checkpoint is considered as a compaction of
+ * a set of journal log entries. If the master does not do any checkpoint, the journal should
+ * still be sufficient.
  *
- * To construct the full state of the master, all the entries must be applied to the checkpoint in
- * order. The entry file most recently being written to is in the base journal folder, where the
- * completed entry files are in the "completed" folder.
+ * Journal file structure:
+ * journal_folder/version/logs/StartSequenceNumber-EndSequenceNumber
+ * journal_folder/version/checkpoints/0-EndSequenceNumber
+ * journal_folder/version/.tmp/random_id
  */
 @ThreadSafe
 public class UfsJournal implements Journal {
-  /** The log number for the first completed log. */
-  protected static final long FIRST_COMPLETED_LOG_NUMBER = 1L;
-  /** The folder for completed logs. */
-  private static final String COMPLETED_LOCATION = "completed";
-  /** The file extension for the current log file. */
-  private static final String CURRENT_LOG_EXTENSION = ".out";
-  /** The file name of the checkpoint file. */
-  private static final String CHECKPOINT_FILENAME = "checkpoint.data";
-  /** The base of the entry log file names, without the file extension. */
-  private static final String ENTRY_LOG_FILENAME_BASE = "log";
+  private static final Logger LOG = LoggerFactory.getLogger(UfsJournal.class);
+  public static final long UNKNOWN_SEQUENCE_NUMBER = Long.MAX_VALUE;
```

Comment:
Should we use a negative number instead (since those are not in the domain of valid sequence numbers)?